File: fix_llm_messages.py
import sys

with open('src/services/llmService.js', 'r') as f:
    content = f.read()

# Replace the mangled _prepareMessages and the old _ensureUserMessage
start_marker = "  _prepareMessages(messages, systemPrompt) {"
end_marker = "  async generateResponse(messages, options = {}) {"

start_idx = content.find(start_marker)
end_idx = content.find(end_marker)

if start_idx != -1 and end_idx != -1:
    new_method = """  _prepareMessages(messages, systemPrompt) {
    const prepared = [];
    if (systemPrompt) {
      prepared.push({ role: 'system', content: systemPrompt });
    }

    // Filter out invalid or empty messages
    const validMessages = (messages || []).filter(m => m && m.content && m.content.trim() !== "");
    prepared.push(...validMessages);

    const hasUser = prepared.some(m => m.role === 'user');
    if (!hasUser) {
      // API requires at least one user message
      prepared.push({ role: 'user', content: 'Proceed.' });
    }

    return prepared;
  }

"""
    content = content[:start_idx] + new_method + content[end_idx:]

# Update the call in generateResponse
content = content.replace(
    "const fullMessages = this._ensureUserMessage([{ role: 'system', content: systemPrompt }, ...messages]);",
    "const fullMessages = this._prepareMessages(messages, systemPrompt);"
)

with open('src/services/llmService.js', 'w') as f:
    f.write(content)

File: fix_llm_messages.js
import fs from 'fs/promises';

async function fix() {
  const path = 'src/services/llmService.js';
  let content = await fs.readFile(path, 'utf-8');

  const newPrepareMessages = `  _prepareMessages(messages, systemPrompt) {
    const prepared = [];

    // Ensure system prompt is first
    if (systemPrompt) {
      prepared.push({ role: 'system', content: systemPrompt });
    }

    // Filter and add other messages
    const otherMessages = (messages || []).filter(m => m && m.content && m.content.trim() !== "");

    const contextSystemMessages = otherMessages.filter(m => m.role === 'system');
    const nonSystemMessages = otherMessages.filter(m => m.role !== 'system');

    if (contextSystemMessages.length > 0) {
      if (prepared.length > 0) {
        // Append context system messages to the main one
        const combinedSystem = prepared[0].content + "\\n\\n" + contextSystemMessages.map(m => m.content).join("\\n");
        prepared[0].content = combinedSystem;
      } else {
        const combinedSystem = contextSystemMessages.map(m => m.content).join("\\n");
        prepared.push({ role: 'system', content: combinedSystem });
      }
    }

    prepared.push(...nonSystemMessages);

    const hasUser = prepared.some(m => m.role === 'user');
    if (!hasUser) {
      prepared.push({ role: 'user', content: 'Proceed.' });
    }

    return prepared;
  }`;

  // Replace _prepareMessages body
  const start = content.indexOf('_prepareMessages(messages, systemPrompt) {');
  const braceStart = content.indexOf('{', start);
  let count = 1;
  let pos = braceStart + 1;
  while (count > 0 && pos < content.length) {
    if (content[pos] === '{') count++;
    else if (content[pos] === '}') count--;
    pos++;
  }

  content = content.slice(0, start) + newPrepareMessages + content.slice(pos);

  await fs.writeFile(path, content);
  console.log('Fixed message ordering in LLMService.');
}
fix();
